File: src/cursor.lua
Cursor = class()

function Cursor:Create(grid)
  self = self:New()
  local mesh = CreateCubeMesh()
  self.entity = CreateModel(mesh)
  self.grid = grid
  self.alpha = 0.5
  self.alpha_dir = 1
  FreeMesh(mesh)
  local mat = EntityMaterial(self.entity, 1)
  SetMaterialType(mat, MATERIAL_ALPHA)
  SetMaterialFlag(mat, FLAG_LIGHTING, false)
  SetMaterialFlag(mat, FLAG_VERTEXCOLORS, true)
  self:reset()
  return self
end

function Cursor:reset()
  SetEntityPosition(self.entity, self.grid:tilesX() / 2, self.grid:tilesY() / 2, self.grid:tilesZ() / 2)
end

function Cursor:update(editing)
  if editing and not EntityVisible(self.entity) then SetEntityVisible(self.entity, true) end
  if not editing and EntityVisible(self.entity) then SetEntityVisible(self.entity, false) end
  if editing then
    -- Update alpha
    self.alpha = self.alpha + self.alpha_dir * 0.5 * DeltaTime()
    if self.alpha <= 0.25 or self.alpha >= 0.75 then
      self.alpha = Clamp(self.alpha, 0.25, 0.75)
      self.alpha_dir = self.alpha_dir * -1
    end
    SetMeshColor(ModelMesh(self.entity), FadeColor(COLOR_ORANGE, self.alpha * 255))
    UpdateMesh(ModelMesh(self.entity))

    -- Update position
    if KeyHit(KEY_UP) then TranslateEntity(self.entity, 0, 0, 1) end
    if KeyHit(KEY_DOWN) then TranslateEntity(self.entity, 0, 0, -1) end
    if KeyHit(KEY_LEFT) then TranslateEntity(self.entity, -1, 0, 0) end
    if KeyHit(KEY_RIGHT) then TranslateEntity(self.entity, 1, 0, 0) end
    if KeyHit(KEY_Q) then TranslateEntity(self.entity, 0, 1, 0) end
    if KeyHit(KEY_A) then TranslateEntity(self.entity, 0, -1, 0) end
    SetEntityPosition(
      self.entity,
      Clamp(EntityX(self.entity), 1, self.grid:tilesX()),
      Clamp(EntityY(self.entity), 1, self.grid:tilesY()),
      Clamp(EntityZ(self.entity), 1, self.grid:tilesZ()))
  end
end

File: src/cursor.cpp
#include "cursor.h"
#include <algorithm>

Cursor::Cursor(Grid& g) : grid(g), alpha(0.5f), alpha_dir(1.f) {
    Mesh* mesh = CreateCubeMesh();
    entity = CreateModel(mesh, nullptr);
    FreeMesh(mesh);

    Material* mat = GetEntityMaterial((Entity*)entity, 0);
    SetMaterialBlendMode(mat, BLEND_ALPHA);
    SetMaterialLightingEnabled(mat, FALSE);
    SetMaterialColor(mat, ChangeAlpha(COLOR_ORANGE, (int)(alpha * 255.f)));
    reset();
}

Cursor::~Cursor() {
    if (entity) FreeModel(entity);
}

void Cursor::reset() {
    App::set_entity_position((Entity*)entity,
        (float)grid.tiles_x() / 2.f,
        (float)grid.tiles_y() / 2.f,
        (float)grid.tiles_z() / 2.f);
}

void Cursor::update(bool editing) {
    SetEntityVisible((Entity*)entity, editing ? TRUE : FALSE);

    if (!editing) return;

    alpha += alpha_dir * 0.5f * App::delta_time;
    if (alpha <= 0.25f || alpha >= 0.75f) {
        alpha     = std::min(std::max(alpha, 0.25f), 0.75f);
        alpha_dir = -alpha_dir;
    }

    Material* mat = GetEntityMaterial((Entity*)entity, 0);
    SetMaterialColor(mat, ChangeAlpha(COLOR_ORANGE, (int)(alpha * 255.f)));

    if (App::key_hit[GLFW_KEY_UP])    App::translate_entity((Entity*)entity, 0.f,  0.f,  1.f);
    if (App::key_hit[GLFW_KEY_DOWN])  App::translate_entity((Entity*)entity, 0.f,  0.f, -1.f);
    if (App::key_hit[GLFW_KEY_LEFT])  App::translate_entity((Entity*)entity, -1.f, 0.f,  0.f);
    if (App::key_hit[GLFW_KEY_RIGHT]) App::translate_entity((Entity*)entity, 1.f,  0.f,  0.f);
    if (App::key_hit[GLFW_KEY_Q])     App::translate_entity((Entity*)entity, 0.f,  1.f,  0.f);
    if (App::key_hit[GLFW_KEY_A])     App::translate_entity((Entity*)entity, 0.f, -1.f,  0.f);

    float cx = std::min(std::max(App::entity_x((Entity*)entity), 1.f), (float)grid.tiles_x());
    float cy = std::min(std::max(App::entity_y((Entity*)entity), 1.f), (float)grid.tiles_y());
    float cz = std::min(std::max(App::entity_z((Entity*)entity), 1.f), (float)grid.tiles_z());
    App::set_entity_position((Entity*)entity, cx, cy, cz);
}
